pages/dev/general-message-passing/cosmos-gmp.mdx

```diff
@@ -0,0 +1,165 @@
+
+
+<Callout emoji="💡">
+To see full examples of sending and receiving messages in all cases, check out our full [Cosmos GMP Example Project](https://github.com/axelarnetwork/evm-cosmos-gmp-sample/).
+</Callout>
+
+## Messages from EVM to CosmWasm
+To send a message from an EVM chain to a Cosmos CosmWasm contract, the payload must be encoded before calling the [standard](./gmp-messages) `payNativeGasForContractCall` and `callContract` methods (or [token-including equivalents](./gmp-tokens-with-messages)). You must also specify the destination method you are calling.
+
+
+The payload must specify:
+
+1. **string** - The method name of the CosmWasm contract to call
+1. **string[]** - The names of all arguments
+1. **string[]** - the types of all arguments
+1. **bytes** - the values of all arguments
+
+### MultiSend Example encoding
+This example encodes a string array of recipients, as you can see in our [example of sending tokens and recipients](https://github.com/axelarnetwork/axelar-examples/blob/main/examples/evm/call-contract-with-token/DistributionExecutable.sol).
+
+```solidity
+bytes memory argValue = abi.encode(recipients); // A standard EVM payload
+bytes memory payload  = abi.encode(
+    "multi_send", // CosmWasm method name
+    StringArray.fromArray1(["recipients"]), // argument name
+    StringArray.fromArray1(["string[]"]), // argument type
+    argValue // argument value
+);
+bytes memory payloadToCW = abi.encodePacked(
+    bytes4(uint32(1)), // version number
+    payload
+);
+```
+
+## Messages from EVM to Native Cosmos
+When sending a message to a Cosmos chain with native (consensus layer) support for GMP, you must encode a version number with the intended payload. This allows the consensus layer to know what to do with the payload.
+
+```solidity
+bytes memory payload = abi.encodePacked(
+    bytes4(uint32(0)), // version number
+    abi.encode(receiverAddresses)
+);
+```
+
+
+## Messages from Native Cosmos
+To send a message from Naive Cosmos Go code, you must send a message to the Axelar GMP Account (`axelar1dv4u5k73pzqrxlzujxg3qp8kvc3pje7jtdvu72npnt5zhq05ejcsn5qme5s`) via IBC.
+
+
+
+When sending messages from Cosmos, you should perform standard GMP ABI encoding via the `github.com/ethereum/go-ethereum/accounts/abi` package. 
+
+```go
+// build payload that can be decoded by solidity
+addressesType, err := abi.NewType("address[]", "address[]", nil)
+if err != nil {
+    return nil, err
+}
+
+var addresses []common.Address
+for _, receiver := range msg.ReceiverAddresses {
+    addresses = append(addresses, common.HexToAddress(receiver))
+}
+
+payload, err := abi.Arguments{{Type: addressesType}}.Pack(addresses)
+if err != nil {
+    return nil, err
+}
+
+message := Message{
+    DestinationChain:   msg.DestinationChain,
+    DestinationAddress: msg.DestinationAddress,
+    Payload:            payload,
+    Type:               TypeGeneralMessageWithToken,
+}
```

Comment:
gas service from cosmos -> evm https://github.com/axelarnetwork/evm-cosmos-gmp-sample/tree/main/native-integration#optional-executor-service-for-cosmos---evm